File: BookSearchObjc/BookCollectionObjc/Database/DatabaseManager/DatabaseManager.m
//
//  CoreDataManager.m
//  BookCollectionObjc
//
//  Created by Milton Palaguachi on 11/15/20.
//

#import "DatabaseManager.h"
#import <sqlite3.h>
@interface DatabaseManager ()
@property (nonatomic, strong) NSString *documetDirectory;
@property (nonatomic, strong) NSString *dbPath;
@property (nonatomic, retain) NSString *databaseFileName;
@property (nonatomic, assign, readwrite) sqlite3 *dbPointer;


- (void)copyDatabaseFileIntoDocumentsDirectoryIfNeeded: (NSString *)destPath;
- (void)createdDatabaseIfNeeded;
- (void)createTableIfRequired;

- (sqlite3 *)openConnection;
- (void)closeConnection;
- (sqlite3_stmt *)prepareStatement: (const char *)sql;

@end

@implementation DatabaseManager


+ (id)sharedManager {
    static DatabaseManager * shared = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        shared = [[self alloc] init];
    });
    return shared;
}

- (id)init {
    
    if (self = [super init]) {
        self.databaseFileName = @"StoredBooksInfo.db";
    }
    return self;
}
- (void)dealloc {
    sqlite3_close(self.dbPointer);
}
- (void)startDatabase {
    [self createdDatabaseIfNeeded];
    self.dbPointer = [self openConnection];
    [self createTableIfRequired];
}

- (void)copyDatabaseFileIntoDocumentsDirectoryIfNeeded: (NSString *)destPath {
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSBundle *mainBundle = [NSBundle mainBundle];
    
    if(![fileManager fileExistsAtPath: destPath]) {
        NSString *sourcePath = [[mainBundle resourcePath]stringByAppendingPathComponent:self.databaseFileName];
        NSError *error;
        [fileManager copyItemAtPath:sourcePath toPath:destPath error:&error];
        if (error) {
            NSLog(@"error coping sourchPath to destinationPath\n - %@", error.localizedDescription);
        }
        
    } else {
        NSLog(@"Path Already Exists: %@", self.dbPath);
    }
    
}
- (void)createdDatabaseIfNeeded  {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    self.documetDirectory = [paths objectAtIndex:0];
    NSString *destinationPath = [self.documetDirectory stringByAppendingPathComponent:self.databaseFileName];
    if (destinationPath) {
        self.dbPath = destinationPath;
        [self copyDatabaseFileIntoDocumentsDirectoryIfNeeded:destinationPath];
    } else {
        NSLog(@"Error Destination Path Failed to create");
    }
    
}

- (void)createTableIfRequired {
    const char *createTableQuery = "CREATE TABLE IF NOT EXISTS BookInfo (Title TEXT, Subtitle TEXT, Authors TEXT, Category TEXT, Publisher TEXT, PublishedDate TEXT, Description TEXT, PageCount INTEGER,  AverageRating REAL, Language TEXT,  Thumbnail TEXT,InfoLink TEXT, PreviewLink TEXT, Id INTEGER NOT NULL UNIQUE, PRIMARY KEY(Id AUTOINCREMENT));";
    sqlite3_stmt * createTableStatement = [self prepareStatement: createTableQuery];
    if (sqlite3_step(createTableStatement) == SQLITE_DONE) {
        NSLog(@"Created table Success");
    } else {
        NSLog(@"Created table Failed");
    }
    sqlite3_finalize(createTableStatement);
}

- (sqlite3 *)openConnection {
    
    sqlite3 *dbPointer;
    if (sqlite3_open([self.dbPath UTF8String], &dbPointer) == SQLITE_OK) {
        return dbPointer;
        
    } else {
        sqlite3_close(dbPointer);
        NSLog(@"error - Failed to open(sqlite3)");
    }
    return NULL;
}
- (void)closeConnection {
    sqlite3_close(self.dbPointer);
    
}
- (sqlite3_stmt *)prepareStatement: (const char *)sql {
    sqlite3_stmt *statement = nil;
    if (sqlite3_prepare_v2(self.dbPointer, sql, -1, &statement, NULL) == SQLITE_OK) {
        NSLog(@"prepare success");
        return statement;
    }
    NSLog(@"prepare Failed");
    return nil;
}


- (void)deleteBook:(int)bookId {
    const char *deleteQuery = "DELETE FROM BookInfo WHERE Id = ?;";
    sqlite3_stmt * deleteStatement = [self prepareStatement:deleteQuery];
    if (sqlite3_bind_int(deleteStatement, 1, bookId) == SQLITE_OK) {
        if (sqlite3_step(deleteStatement) == SQLITE_DONE) {
            NSLog(@"delete success");
            
        } else {
            NSLog(@"delete Failed");
        }
    } else {
        NSLog(@"Error bind to delete row");
    }
    sqlite3_finalize(deleteStatement);
}

- (void)insert:(nonnull Book *)book {
    const char *insertQuery = "INSERT INTO BookInfo (Title, Subtitle, Authors, Category, Publisher, PublishedDate, Description, PageCount, AverageRating, Language, Thumbnail ,InfoLink, PreviewLink) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
    
    sqlite3_stmt * insertStatement = [self prepareStatement: insertQuery];
    NSLog(@"book title: %@", book.title);
    
    if (sqlite3_bind_text(insertStatement, 1, [book.title UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 2, [book.subtitle UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 3, [book.author UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 4, [book.category UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 5, [book.publisher UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 6, [book.publishedDate UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 7, [book.descriptionInfo UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 8, [book.pageCount UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 9, [book.averageRating UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 10, [book.language UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 11, [book.thumbnail UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 12, [book.infoLink UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK && sqlite3_bind_text(insertStatement, 13, [book.previewLink UTF8String], -1, SQLITE_TRANSIENT) == SQLITE_OK) {
        
        if (sqlite3_step(insertStatement) == SQLITE_DONE) {
            NSLog(@"insert in bd Success");
        } else {
            NSLog(@"Failed insert to db");
        }
    }
    else {
        NSLog(@"Faided to bind to aqlite");
        
    }
    sqlite3_finalize(insertStatement);
}

- (nonnull NSArray<Book *> *)readBooks {
    const char * readQuery = "SELECT *FROM BookInfo;";
    sqlite3_stmt *readStatement = [self prepareStatement:readQuery];
    NSMutableArray<Book *> *tempBooks = [NSMutableArray new];
    while(sqlite3_step(readStatement) == SQLITE_ROW) {
        char *titleChars = (char *)sqlite3_column_text(readStatement, 0);
        char *subtitleChars = (char *)sqlite3_column_text(readStatement, 1);
        char *authorChars = (char *)sqlite3_column_text(readStatement, 2);
        char *categoryChars = (char *)sqlite3_column_text(readStatement, 3);
        char *publisherChars = (char *)sqlite3_column_text(readStatement, 4);
        char *publishedDateChars = (char *)sqlite3_column_text(readStatement, 5);
        char *descriptionChars = (char *)sqlite3_column_text(readStatement, 6);
        char *pageCountChars = (char *)sqlite3_column_text(readStatement, 7);
        char *avgRatingChars = (char *)sqlite3_column_text(readStatement, 8);
        char *languageChars = (char *)sqlite3_column_text(readStatement, 9);
        char *thumbnailChars = (char *)sqlite3_column_text(readStatement, 10);
        char *infoLinkChars = (char *)sqlite3_column_text(readStatement, 11);
        char *previewLinkChars = (char *)sqlite3_column_text(readStatement, 12);
        int uniqueId = sqlite3_column_int(readStatement, 0);
        
        NSString *title = [[NSString alloc] initWithUTF8String:titleChars];
        NSString *subtitle = [[NSString alloc] initWithUTF8String:subtitleChars];
        NSString *author = [[NSString alloc] initWithUTF8String:authorChars];
        NSString *category = [[NSString alloc] initWithUTF8String:categoryChars];
        NSString *publisher = [[NSString alloc] initWithUTF8String:publisherChars];
        NSString *publishedDate = [[NSString alloc] initWithUTF8String:publishedDateChars];
        NSString *description = [[NSString alloc] initWithUTF8String:descriptionChars];
        NSString *pageCount = [[NSString alloc] initWithUTF8String:pageCountChars];
        NSString *avgRating = [[NSString alloc] initWithUTF8String:avgRatingChars];
        NSString *language = [[NSString alloc] initWithUTF8String:languageChars];
        NSString *thumbnail = [[NSString alloc] initWithUTF8String:thumbnailChars];
        NSString *infoLink = [[NSString alloc] initWithUTF8String:infoLinkChars];
        NSString *previewLink = [[NSString alloc] initWithUTF8String:previewLinkChars];
        
        Book *book = [Book alloc];
        book.title = title;
        book.subtitle = subtitle;
        book.author = author;
        book.category = category;
        book.publisher = publisher;
        book.publishedDate = publishedDate;
        book.descriptionInfo = description;
        book.pageCount = pageCount;
        book.averageRating = avgRating;
        book.language = language;
        book.thumbnail = thumbnail;
        book.infoLink = infoLink;
        book.previewLink = previewLink;
        book.uniqueId = uniqueId;

        [tempBooks addObject:book];

    }
    sqlite3_finalize(readStatement);
    return tempBooks;
}

- (void)updateBook:(nonnull Book *)book bookId: (int)bookId {
}

@end

File: BookSearchSwift/BookCollection/Database/DatabaseManager/DatabaseManager.swift
//
//  DatabaseManager.swift
//  BookCollection
//
//  Created by Milton Palaguachi on 11/17/20.
//  Copyright © 2020 Milton. All rights reserved.
//

import SQLite3
import Foundation

// MARK: - Custom Enum of Type Error
enum SqliteError: Error {
    
    case invalidDirecotyUrl
    case openDBFailed
    case prepareFailed
    case stepFailed
    case bindFailed
    case invalidBundleUrl
    case copyFailed
    case tableCreationFailed
    case insertSQLiteFailed
    case deleteFailed
    case updateFailed
    case readFailed
    case deleteAllFailed
}

// MARK: - Protocol for Data Provider
protocol DatabaseProvider {
    func insert(volume: VolumeInfoModel) throws
    func delete(whereId pKey: Int32) throws
    func update(volume: VolumeInfoModel, whereId pKey: Int) throws
    func read() throws -> [VolumeInfoModel]
    func insertUsers(volumes: [VolumeInfoModel]) throws
    
    init(dbName: String)
    
    var dbPath: String { get }
    var dbName: String { get }
}

// MARK: - Data Base Handler class created
class DBManager: DatabaseProvider {
   
    var dbPath: String = ""
    var dbName: String
    var dbPointer: OpaquePointer?
    static let shared = DBManager()
    
    // MARK: - Construction(Init)
    internal required init(dbName: String = "StoredBooksInfo.db") {
        self.dbName = dbName
    }
    
    func startDB() {
        try? self.createDBIfRequired()
        self.dbPointer = try? self.openConnection()
        try? createTableIfRequired()
    }
    
    // MARK: - Create Dadat Base IF Required
    private func createDBIfRequired() throws {
        // Get the docs directory
        guard let docDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else { throw SqliteError.invalidDirecotyUrl }
        self.dbPath = docDirectory.appendingPathComponent(self.dbName).path
        do {
            try self.copyDatabaseIntoDocumetsDirectoryIfNeeded(addDestination: self.dbPath)
        } catch let error as SqliteError {
            throw error
        }
    }
    
    // MARK: - Copy Data Base If Needed
    func copyDatabaseIntoDocumetsDirectoryIfNeeded(addDestination destPath: String) throws {
        guard let bundleUrl = Bundle.main.resourceURL?.appendingPathComponent(self.dbName) else {
            throw SqliteError.invalidBundleUrl
        }
        if FileManager.default.fileExists(atPath: destPath) {
            print("Document File Already Exist")
            print(destPath)
        } else if FileManager.default.fileExists(atPath: bundleUrl.path) {
            do {
                print(destPath)
                try FileManager.default.copyItem(atPath: bundleUrl.path, toPath: destPath)
            } catch {
                throw SqliteError.copyFailed
            }
        }
    }
    
    // MARK: - Open DB Connection
    func openConnection() throws -> OpaquePointer? {
        // opquePointer is required to access the db or interact
        var opaquePointer: OpaquePointer? // It's Swift type for C pointer.
        if sqlite3_open(self.dbPath, &opaquePointer) == SQLITE_OK {
            print("Success Open Connection DB")
            return opaquePointer
            
        } else {
            print("Faild to open connection DB")
            throw SqliteError.openDBFailed
        }
    }
    
    // MARK: - Close DB Connection
    func closeConnection() {
        sqlite3_close(self.dbPointer)
    }
    
    // MARK: - Create DB Table If does not exist
    func createTableIfRequired() throws {
                
     let createTableQuery = """
        CREATE TABLE IF NOT EXISTS "BookInfo" (
                    "Title" TEXT,
                    "Authors" TEXT,
                    "Subtitle" TEXT,
                    "Page Count" INTEGER,
                    "Publisher" TEXT,
                    "Published Date" TEXT,
                    "Description" TEXT,
                    "Average Rating" REAL,
                    "Ratings Count" INTEGER,
                    "Language" TEXT,
                    "Preview Link" TEXT,
                    "Book Image Url" TEXT,
                    "Id" INTEGER NOT NULL UNIQUE,
                PRIMARY KEY("Id" AUTOINCREMENT)
             );
        """

        let dbHandler = try self.prepareStatement(sql: createTableQuery)
        
        defer {
            // Finalize deletes the compiled statement to avoid memory licked
            sqlite3_finalize(dbHandler)
        }
        guard sqlite3_step(dbHandler) == SQLITE_DONE else {
            print("Table no Created\n**Throw error**")
            throw SqliteError.tableCreationFailed
        }
        print("Created table Success")
    }
    
    // MARK: - Prepare DB
    func prepareStatement(sql: String) throws -> OpaquePointer? {
        var localPointer: OpaquePointer?
        if sqlite3_prepare_v2(self.dbPointer, sql, -1, &localPointer, nil) == SQLITE_OK {
            return localPointer
            
        } else {
            throw SqliteError.prepareFailed
        }
    }
    
    // MARK: - Destructor(deinit)
    deinit {
        sqlite3_close(self.dbPointer)
    }
}

extension DBManager {
    
    func insertUsers(volumes: [VolumeInfoModel]) throws {
        for  volume in volumes {
            try self.insert(volume: volume)
        }
    }

    // MARK: - Insert UserInfo Row into DB Table
    func insert(volume: VolumeInfoModel) throws {
        let query = """
        INSERT INTO BookInfo (  Title,
                                Authors,
                                Subtitle,
                                `Page Count`,
                                Publisher,
                                `Published Date`,
                                Description,
                                `Average Rating`,
                                `Ratings Count`,
                                Language,
                                `Preview Link`,
                                `Book Image Url`)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
        """

        let statement = try self.prepareStatement(sql: query)
        defer {
            sqlite3_finalize(statement)
        }
        guard sqlite3_bind_text(statement, 1, ((volume.title ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK &&
              sqlite3_bind_text(statement, 2, ((volume.authors ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK &&
                sqlite3_bind_text(statement, 3, ((volume.subtitle ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK &&
                sqlite3_bind_int(statement, 4, Int32(volume.pageCount ?? 0)) == SQLITE_OK &&
                sqlite3_bind_text(statement, 5, ((volume.publisher ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK &&
                sqlite3_bind_text(statement, 6, ((volume.publishedDate ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK &&
                sqlite3_bind_text(statement, 7, ((volume.description ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK &&
                sqlite3_bind_double(statement, 8, volume.averageRating ?? 0.0 ) == SQLITE_OK &&
                sqlite3_bind_int(statement, 9, Int32(volume.ratingsCount ?? 0)) == SQLITE_OK &&
                sqlite3_bind_text(statement, 10, ((volume.language ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK &&
                sqlite3_bind_text(statement, 11, ((volume.previewLink ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK &&
                sqlite3_bind_text(statement, 12, ((volume.bookImageUrl ?? "") as NSString).utf8String, -1, nil) == SQLITE_OK else {
            throw SqliteError.bindFailed
        }
        guard sqlite3_step(statement) == SQLITE_DONE else {
            throw SqliteError.insertSQLiteFailed
        }
        print("Values written Successfully")
    }
    // MARK: - Delete All Rows in DB Table
    func deleteAll() throws {
        let query = "DELETE FROM BookInfo"
        let statement = try self.prepareStatement(sql: query)
        defer {
            sqlite3_finalize(statement)
        }
        guard sqlite3_step(statement) == SQLITE_DONE else {
            throw SqliteError.deleteAllFailed
        }
        print("All Value deleted Succeessfully")
    }
    
    // MARK: - Delete A Row By Id in Table
    func delete(whereId pKey: Int32) throws {
        let query = "DELETE FROM BookInfo WHERE Id = ?;"
        let statement = try self.prepareStatement(sql: query)
        guard sqlite3_bind_int(statement, 1, pKey) == SQLITE_OK else { throw SqliteError.bindFailed }
        
        defer {
            sqlite3_finalize(statement)
            print("Deleted volume with id: \(pKey)")
        }
        guard sqlite3_step(statement) == SQLITE_DONE else {
            throw SqliteError.deleteFailed
        }
        print("Value deleted Succeessfully")
    }
    
    // MARK: - Update Rows in Table
    func update(volume: VolumeInfoModel, whereId pKey: Int) throws {
  
        let query = """
        UPDATE BookInfo
        SET `Title` = '\(volume.title ?? "")',
            `Authors` = '\(volume.authors ?? "")',
            `Subtitle` = '\(volume.subtitle ?? "")',
            `Page Count` = '\(volume.pageCount ?? 0)',
            `Publisher` = '\(volume.publisher ?? "")',
            `Published Date` = '\(volume.publishedDate ?? "")',
            `Description` = '\(volume.description ?? "")',
            `Averate Rating` '\(volume.averageRating ?? 0.0)',
            `Ratings Count` = '\(volume.ratingsCount ?? 0)',
            `Language` = '\(volume.language ?? "")',
            `Preview Link` = '\(volume.previewLink ?? "")',
            `Book Image Url` = '\(volume.previewLink ?? "")'

        WHERE Id = \(pKey)
        """

        let statement = try self.prepareStatement(sql: query)
        defer {
            sqlite3_finalize(statement)
        }
        guard sqlite3_step(statement) == SQLITE_DONE else {
            throw SqliteError.updateFailed
        }
        print("updated Value Succeessfully")
    }
    
    // MARK: - Read Al Rows From Table
    func read() throws -> [VolumeInfoModel] {
        print("read")
        let query = "SELECT * FROM BookInfo;"
        let statement = try self.prepareStatement(sql: query)
        var volumes = [VolumeInfoModel]()
        while sqlite3_step(statement) == SQLITE_ROW {
            print("reading rows")
            let title = String(describing: String(cString: sqlite3_column_text(statement, 0)))
            let authors = String(describing: String(cString: sqlite3_column_text(statement, 1)))
            let subtitle = String(describing: String(cString: sqlite3_column_text(statement, 2)))
            
            let pageCount = Int(sqlite3_column_int(statement, 3))
            
            let publisher = String(describing: String(cString: sqlite3_column_text(statement, 4)))
            let publishedDate = String(describing: String(cString: sqlite3_column_text(statement, 5)))
            let description = String(describing: String(cString: sqlite3_column_text(statement, 6)))
            
            let avgRating = sqlite3_column_double(statement, 7)
            let ratingCount = Int(sqlite3_column_int(statement, 8))
            
            let language = String(describing: String(cString: sqlite3_column_text(statement, 9)))
            let priviewLink = String(describing: String(cString: sqlite3_column_text(statement, 10)))
            let bookImageUrl = String(describing: String(cString: sqlite3_column_text(statement, 11)))

            let pId = sqlite3_column_int(statement, 12)
            var volume = VolumeInfoModel(volume: nil)
            volume.identifier = pId
            volume.authors = authors
            volume.title = title
            volume.subtitle = subtitle
            volume.pageCount = pageCount
            volume.publisher = publisher
            volume.publishedDate = publishedDate
            volume.description = description
            volume.averageRating = avgRating
            volume.ratingsCount = ratingCount
            volume.language = language
            volume.previewLink = priviewLink
            volume.bookImageUrl = bookImageUrl
            volumes.append(volume)
        }
        print("read Data Successfully")
        return volumes
    }
}
